### ISSUE TITLE
Keys for closed doors don't blink with regular HUD

### ISSUE THREAD

**Issue by Zodomaniac** [fabiangreffrath/crispy-doom#382]
### Background

Version of Crispy Doom: Daily 25.01.19

Operating System and version: Win 7

Game: (Doom/Heretic/Hexen/Strife/other) Doom (2)

Any loaded WADs and mods (please include full command line):

### Bug description

Observed behavior: When attempting to open the door that I have no key for, the key blinks in the HUD only if I have Crispy HUD enabled.

**Comment by fabiangreffrath** [comment]
This is intentional. The Crispy HUD is already more feature-rich (e.g. gibbed and crushed player corpse in the health widget, berserk pack in the ammo widget) whereas the classic status bar is strictly Vanilla.

The Crispy HUD is "my invention", so I feel more free to add features, but I am really not sure if this should be enabled for the classic status bar as well.

**Comment by Zodomaniac** [comment]
I remember there was some issue with blinking keys back in the day, so I thought you refused from blinking keys in classic HUD for some bug reason.

**Comment by fabiangreffrath** [comment]
Yes, there is some bug in the commented out code that would enable this for the classic status bar, but this is not the reason why it's still disabled there.

**Comment by Zodomaniac** [comment]
I can't remember what exactly kind of bug that was and which thread it was discussed in.

**Comment by JNechaevsky** [comment]
https://github.com/fabiangreffrath/crispy-doom/commit/6e1bff71914f897227aab2f57b1b00578baedc18#commitcomment-31775337

I wonder if it possible to convert key's backgound to something like numbers background. Probably this way it will not be necessary to make a full redraw every tic. Nevermind, I'll try to investigate later.

**Comment by fabiangreffrath** [comment]
This could be it:
```
--- a/src/doom/st_stuff.c
+++ b/src/doom/st_stuff.c
@@ -1435,7 +1435,7 @@ void ST_updateWidgets(void)
                }
 #endif
 #if defined(CRISPY_KEYBLINK_IN_CLASSIC_HUD)
-               if (screenblocks < CRISPY_HUD && !(plyr->tryopen[i] & (KEYBLINKMASK-1)))
+               if (screenblocks < CRISPY_HUD && (!(plyr->tryopen[i] & (KEYBLINKMASK-1)) || plyr->tryopen[i] == 1))
                {
                        st_firsttime = true;
                }
```

**Comment by JNechaevsky** [comment]
Ooooh, our department was celebrating Student's Day, I'm bloated by pizza and champagne and lost a gaming party in table soccer to my Chief with 2:7 score. Life is going well. Help me. :sweat_smile:

Will check it in few hours, need to get back to the human condition first.

**Comment by JNechaevsky** [comment]
Nope, still same problem with:

```
+ #define CRISPY_KEYBLINK_IN_CLASSIC_HUD

#if defined(CRISPY_KEYBLINK_IN_CLASSIC_HUD)
-               if (screenblocks < CRISPY_HUD && !(plyr->tryopen[i] & (KEYBLINKMASK-1)))
+               if (screenblocks < CRISPY_HUD && (!(plyr->tryopen[i] & (KEYBLINKMASK-1)) || plyr->tryopen[i] == 1))
		{
			st_firsttime = true;
		}
```

Picked up key is not appearing after blinking sequence was finished until full status bar update. I still have bad suspection, that full update must be done at least for one tic right _after_ blinking sequence was finished... But I don't see any catchers for such `+1 tic` code.

**Comment by fabiangreffrath** [comment]
Need to rethink the timing in this code - although I don't have plans to use it myself...

**Comment by JNechaevsky** [comment]
Leave it to me please, I still have a basic, raw idea to update keys background just like it is done with numbers.

**Comment by fabiangreffrath** [comment]
Does this help?
```
--- a/src/doom/st_stuff.c
+++ b/src/doom/st_stuff.c
@@ -1447,6 +1447,11 @@ void ST_updateWidgets(void)
                {
                        keyboxes[i] = (plyr->tryopen[i] & KEYBLINKMASK) ? i + st_keyorskull[i] : -1;
                }
+
+               if (!plyr->tryopen[i])
+               {
+                       w_keyboxes[i].oldinum = -1;
+               }
        }
     }

```

**Comment by JNechaevsky** [comment]
Oh, by the way, about the Google table you have sent me - this gives me an unusual inspiration for creativity. I'll explain privately, the thing that was happened to me was something definitely new.

**Comment by fabiangreffrath** [comment]
Yep, this is exactly how it's supposed to look. The culprit was that the key widget only refreshes itself, if its current state is different than its previous state. When we set its current state to `-1` after the last blink, its previous state is `1`. So, when you pick up the key, its current state changes to `1`, but this is not different from its previous state then. Thus, the widget does not refresh itself.

By changing its previous state back to `-1` after the last blink, picking up the key triggers a refresh of the widget.

**Comment by fabiangreffrath** [comment]
> the Google table you have sent me

The table was meant to be parametric and formatted, I have fixed this now.
